### ISSUE TITLE
Cluster crashed after adding a node while another node is down

### ISSUE THREAD

**Issue by tinco** [canonical/microk8s#2348]
[inspection-report-20210618_165740.tar.gz](https://github.com/ubuntu/microk8s/files/6678391/inspection-report-20210618_165740.tar.gz)

We have a 3 node cluster, one of the nodes had a hardware failure last week, and today I replaced it with a new node. After setting it up, enabling gpu, I ran add-node on the main node, and then join on the new node. I should probably have ran `remove-node --force` for the missing node, but I forgot that step. 

To my horror after running the join command the whole cluster went unavailable. `microk8s status` keeps returning:

```
microk8s is not running. Use microk8s inspect for a deeper inspection.
```

I can't curl the api server either, so I think it's not running at all.

```
tinco@automator-1:~$ curl https://127.0.0.1:16443
curl: (7) Failed to connect to 127.0.0.1 port 16443: Connection refused
```

I can't figure out what exactly is preventing the api server from running, there's no errors in the api-server logs that I can see.

**Comment by tinco** [comment]
Alright, have some more info, pretty sure it's the dqlite cluster that's broken, because this command just hangs forever:

```
tinco@automator-1:~/inspection-report$ microk8s dbctl backup
Backing up the datastore
INFO[0000] mode: backup-dqlite, endpoint: unix:///var/snap/microk8s/current/var/kubernetes/backend/kine.sock, dir: /tmp/tmpvaf2by57/backup-2021-06-18-21-01-04
```

**Comment by balchua** [comment]
I think dqlite didn't start on time.  

Did you try to check the cluster from the other 2 nodes?
It will also be interesting to see whats in your `/var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml`

**Comment by tinco** [comment]
Thanks!

```
- Address: 192.168.1.128:19001
  ID: 3297041220608546238
  Role: 0
- Address: 192.168.1.178:19001
  ID: 13453059637000826160
  Role: 0
- Address: 192.168.1.231:19001
  ID: 3236320156238981192
  Role: 0
```

So the broken node that's gone is "192.168.1.231", and the new node that broke the cluster upon joining isn't in this list (it's 192.168.1.36).

Most important for me is bringing at least 192.168.1.128 back up, with its original configuration. Could I remove the cluster.yaml to achieve this?

**Comment by tinco** [comment]
Oh actually, this is interesting, the cluster.yaml of `192.168.1.36` and of `192.168.1.178` do contain `192.168.1.36` as a member. So it's just `192.168.1.128` that doesn't have the new node in its cluster config.

**Comment by tinco** [comment]
Is there any way to recover the database? All commands I can use seem to have to connect to the dqlite database through the socket, which just doesn't seem to work because I guess it doesn't have a leader or something?

**Comment by tinco** [comment]
I don't even know what node is supposed to be the leader at this point.

**Comment by balchua** [comment]
Hi @tinco you may get some luck trying this.

https://discuss.kubernetes.io/t/recovery-of-ha-microk8s-clusters/12931

**Comment by tinco** [comment]
Hi @balchua ! Thanks a lot for your support. I had actually recovered the cluster at the moment I typed that cluster, I just didn't know it yet. For other people, the link @balchua gave is probably the way to go, but this is what I did:

1. stop all nodes
2. make sure all cluster.yaml files are the same and have the nodes you want in them
3. identify a node that was last healthy, and scp the files that look like this `00000xxxx-000xxxx` to all the other nodes
4. start all nodes
5. *wait 10 minutes*

This got me back all nodes, and the database was still fine.

So I'm rescued now, and can finally start my weekend, but I'm still a bit shocked that the simple process of replacing a failed node cost my team a days work and me a good portion of my weekend. 

Besides being apparently being prone to failure, it is also very much undocumented how to recover from failure. I had googled and I consider myself to be pretty good at googling, and I had not found the link @balchua just posted. Copying over those files was a complete guess on my part, I hadn't even found the location of the database and cluster configuration yet until @balchua gave me the path (/var/snap/microk8s/current/var/kubernetes/backend).

The microk8s website only seems to document the happy path.
